src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -427,7 +427,10 @@ int32_t SystemNative_ReadDirR(DIR* dir, uint8_t* buffer, int32_t bufferSize, Dir
         return errno == 0 ? -1 : errno;
     }
 #else
-    int error = readdir_r(dir, entry, &result);
+    int error;
+
+    // EINTR isn't documented, happens in practice on macOS.
+    while ((error = readdir_r(dir, entry, &result)) && error == EINTR);
```

Comment:
are you sure the EINTR is returned as the return value and not in `errno`?

Reply:
I don't have a reliable repro. I can adapt it so it checks errno, or look through xnu source.

Reply:
Seems like readdir_r sets the return value to whatever is in errno?

Reply:
I've read a few man pages of different OSes and there seems to be little consistency regarding this.
If using `errno` works on macOS then I'd prefer that since it's what we use everywhere else.

Reply:
I've changed it.

Reply:
> If using errno works on macOS then I'd prefer that since it's what we use everywhere else.

Except this API explicitly says errors are returned rather than surfacing through errno.  Is there an example man page that says readdir_r sets errno?  What happens if a readdir_r implementation doesn't touch errno at all, errno retains an EINTR value from prior to the readdir_r call, and readdir_r hits another error... might we end up in an infinite loop?

Reply:
I can change the code back to checking the return code :D

Reply:
Before you do, I'd like to hear from @akoeplinger  :)  I was actually asking questions rather than stating things that must happen ;-)

Reply:
Specifically, on macOS/FreeBSD, it looks like this: https://opensource.apple.com/source/Libc/Libc-1439.40.11/gen/FreeBSD/readdir.c.auto.html